knora/excel2xml.py

```diff
@@ -0,0 +1,2127 @@
+
+
+    if iso_date:
+        year = int(iso_date.group(1))
+        month = int(iso_date.group(2))
+        day = int(iso_date.group(3))
+        try:
+            startdate = datetime.date(year, month, day)
+            enddate = startdate
+        except ValueError:
+            handle_warnings(f"Date parsing error in resource {calling_resource}: '{iso_date.group(0)}' is "
+                            f"not a valid date")
+            return None
+
+    elif eur_date_range:
+        startday = int(eur_date_range.group(1))
+        startmonth = int(eur_date_range.group(2)) if eur_date_range.group(2) else int(eur_date_range.group(5))
+        startyear = int(eur_date_range.group(3)) if eur_date_range.group(3) else int(eur_date_range.group(6))
+        endday = int(eur_date_range.group(4))
+        endmonth = int(eur_date_range.group(5))
+        endyear = int(eur_date_range.group(6))
+        try:
+            startdate = datetime.date(startyear, startmonth, startday)
+            enddate = datetime.date(endyear, endmonth, endday)
+            if enddate < startdate:
+                raise ValueError
+        except ValueError:
+            handle_warnings(f"Date parsing error in resource {calling_resource}: '{eur_date_range.group(0)}' is not a "
+                            f"valid date", stacklevel=3)
+            return None
+
+    elif eur_date:
+        startday = int(eur_date.group(1))
+        startmonth = int(eur_date.group(2))
+        startyear = int(eur_date.group(3))
+        try:
+            startdate = datetime.date(startyear, startmonth, startday)
+            enddate = startdate
+        except ValueError:
+            handle_warnings(f"Date parsing error in resource {calling_resource}: '{eur_date.group(0)}' is not a valid "
+                            f"date", stacklevel=3)
+            return None
+
+    elif monthname_date:
+        day = int(monthname_date.group(2))
+        month = monthes_dict[monthname_date.group(1)]
+        year = int(monthname_date.group(3))
+        try:
+            startdate = datetime.date(year, month, day)
+            enddate = startdate
+        except ValueError:
+            handle_warnings(f"Date parsing error in resource {calling_resource}: '{monthname_date.group(0)}' is not a "
+                            f"valid date", stacklevel=3)
+            return None
+
+    elif year_range:
+        startyear = int(year_range.group(1))
+        endyear = int(year_range.group(2))
+        if int(endyear/100) == 0:
+            # endyear is only 2-digit: add the first two digits of startyear
+            endyear = int(startyear/100) * 100 + endyear
+
+    elif year_only:
+        startyear = int(year_only.group(0))
+        endyear = startyear
+
+    if startdate and enddate:
+        return f"GREGORIAN:CE:{startdate.isoformat()}:CE:{enddate.isoformat()}"
+    elif startyear and endyear:
+        return f"GREGORIAN:CE:{startyear}:CE:{endyear}"
+    else:
+        return None
```

Comment:
Yes, it was a lot of work. This method covers the use cases that had come across in our daily work in the Research Data Unit (formerly Client Services) over the last few monthes. So, this method grew over time.

 1. I have checked `dateparser`, and it would only do a part of the job, plus it has false positives.
 2. I experimented with splitting this up into multiple files. My try was to do this:
```
excel2xml
├── __init__.py
├── excel2xml.py
├── helper_functions.py
└── make_xml_element_functions.py
```
But there were 2 problems:
 1. In the above case, I splitted the methods into 3 groups that logically belong together, so that none of the 3 files was too long. But then I get circular import errors that I cannot resolve. The methods are too much entagled into each other.
 2. It seems odd to me to put only 1 method into a separate file, and leave everything else.

So I didn't succeed in splitting this file up. Have you got ideas what to do? Or, is it really so problematic to have a 2k+ lines long file? I don't see a problem in it...

Reply:
ad 1.:  
with a quick search I found https://datefinder.readthedocs.io/en/latest/ and https://dateutil.readthedocs.io/en/stable/parser.html (from here: https://stackoverflow.com/questions/3276180/extracting-date-from-a-string-in-python). 
But I never used either of those, so I can't say if they're useful at all.

ad 2.:  
the fact that the methods are entangled and you get circular imports in and by itself indicates that the coupling is too high and it should be improved... but I'd have to give it more thought before I can say anything on how it should be done.  
I'd leave the call up to you: I can look into it, or we can decide that it's not a priority. Whatever you prefer is fine by me.

Reply:
(2k+ lines in and by itself is not a problem. But coupling is bad software design - and usually one has long files because one can't split it because the coupling is too high...)